lib/models.dart

```diff
@@ -94,4 +104,25 @@ class BikeState with _$BikeState {
   List<int> toWriteData() {
     return [0, 209, light ? 1 : 0, assist, _modeToWrite(), 0, 0, 0, 0, 0];
   }
+
+  calculateBattery(List<int> data2) {
+    int range = data2[8];
+
+    // Convert the range to %
+    double batteryPercentage = range / 60.0 * 100;
+    batteryPercentage = double.parse(batteryPercentage.toStringAsFixed(1));
+
+    //print('$batteryPercentage %');
+    return batteryPercentage;
+    }
+
+  calculateSpeed(List<int> data3) {
+    int speed = data3[2] + (data3[3] * 256);
+
+    // Convert the speed to km/h
+    double speedKmH = speed / 100.0;
+
+    //print('$speedKmH km/h');
+    return speedKmH;
+    }
```

Comment:
I wonder if it makes more sense to keep all the conversion logic in one place? Right now, the speed data is stored in km here, then converted to mi in the widget. Same with battery and percentage/voltage. I could see storing the raw data in BikeState as we get it from the bike, then have getter methods on BikeState that does the conversion as it's read. Like `speedKM`, `speedMI`, `batteryVolt`, `batteryPercentage`. This has the added benefit of not losing accuracy by not having to double convert.

Reply:
I agree, I will do some magic to make the conversion logic seperated.

`int speed = data3[2] + (data3[3] * 256);`
returns meter per hour.

Reply:
https://github.com/Bamuel/superduper/commit/84b63c67fd4e80033d5bc7373c53dfeff1339431